channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1239 @@
+
+		s = rdpgfx_server_packet_new(
+		    RDPGFX_CMDID_WIRETOSURFACE_1, 
+		    sizeof(RDPGFX_WIRE_TO_SURFACE_PDU_1) + h264Size);
+	}
+	else if (cmd->codecId == RDPGFX_CODECID_AVC444)
+	{
+		havc444 = (RDPGFX_AVC444_BITMAP_STREAM *)cmd->extra;
+		h264Size = sizeof(UINT32); /* cbAvc420EncodedBitstream1 */
+
+		/* avc420EncodedBitstream1 */
+		havc420 = &(havc444->bitstream[0]);
+		h264Size += rdpgfx_estimate_h264_avc420(havc420);
+
+		/* avc420EncodedBitstream2 */
+		if (havc444->LC == 0)
+		{
+			havc420 = &(havc444->bitstream[1]);
+			h264Size += rdpgfx_estimate_h264_avc420(havc420);
+		}
+
+		s = rdpgfx_server_packet_new(
+		    RDPGFX_CMDID_WIRETOSURFACE_1, 
+		    sizeof(RDPGFX_WIRE_TO_SURFACE_PDU_1) + h264Size);
+	}
+	else
+	{
+		s = rdpgfx_server_packet_new(
+		    RDPGFX_CMDID_WIRETOSURFACE_1, 
+		    sizeof(RDPGFX_WIRE_TO_SURFACE_PDU_1) + sizeof(cmd->length));
+	}
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rdpgfx_server_packet_new failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	/* Reformat and send */
+	if (cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE ||
+	    cmd->codecId == RDPGFX_CODECID_CAPROGRESSIVE_V2)
+	{
+		Stream_Write_UINT16(s, cmd->surfaceId); /* surfaceId (2 bytes) */
+		Stream_Write_UINT16(s, cmd->codecId); /* codecId (2 bytes) */
+		Stream_Write_UINT32(s, cmd->contextId); /* codecContextId (4 bytes) */
+		Stream_Write_UINT8(s, cmd->format); /* pixelFormat (1 byte) */
+
+		Stream_Write_UINT32(s, cmd->length); /* bitmapDataLength (4 bytes) */
+		Stream_Write(s, cmd->data, cmd->length);
+
+		return rdpgfx_server_packet_send(context, s);
+	}
+	else
+	{
+		Stream_Write_UINT16(s, cmd->surfaceId); /* surfaceId (2 bytes) */
+		Stream_Write_UINT16(s, cmd->codecId); /* codecId (2 bytes) */
+		Stream_Write_UINT8(s, cmd->format); /* pixelFormat (1 byte) */
+
+		Stream_Write_UINT16(s, cmd->left); /* left (2 bytes) */
+		Stream_Write_UINT16(s, cmd->top); /* top (2 bytes) */
+		Stream_Write_UINT16(s, cmd->right); /* right (2 bytes) */
+		Stream_Write_UINT16(s, cmd->bottom); /* bottom (2 bytes) */
+
+		Stream_Write_UINT32(s, cmd->length); /* bitmapDataLength (4 bytes) */
+		bitmapDataStart = Stream_GetPosition(s);
+
+		if (cmd->codecId == RDPGFX_CODECID_AVC420)
+		{
+			havc420 = (RDPGFX_AVC420_BITMAP_STREAM *)cmd->extra;
+			rdpgfx_write_h264_avc420(context, s, havc420);
+		}
+		else if (cmd->codecId == RDPGFX_CODECID_AVC444)
+		{
+			havc444 = (RDPGFX_AVC444_BITMAP_STREAM *)cmd->extra;
+			havc420 = &(havc444->bitstream[0]);
+
+			/* avc420EncodedBitstreamInfo (4 bytes) */
+			Stream_Write_UINT32(s, havc420->length | (havc444->LC << 30UL));
+
+			/* avc420EncodedBitstream1 */
+			rdpgfx_write_h264_avc420(context, s, havc420);
+
+			/* avc420EncodedBitstream2 */
+			if (havc444->LC == 0)
+			{
+				havc420 = &(havc444->bitstream[0]);
+				rdpgfx_write_h264_avc420(context, s, havc420);
+			}
+		}
+		else
+		{
+			Stream_Write(s, cmd->data, cmd->length);
+		}
+
+		assert(Stream_GetPosition(s) <= Stream_Capacity(s));
```

Comment:
Just want to have an insurance if future we break the stream length estimation. Currently the length is ensured